src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -1360,6 +1463,110 @@ private Object writeReplace() {
+        @Stable
+        private final Map<K, StableValueImpl<V>> delegate;
+
+        LazyMap(Set<K> keys, Function<? super K, ? extends V> mapper) {
+            this.mapper = mapper;
+            this.delegate = StableValueImpl.ofMap(keys);
+        }
+
+        @Override public boolean              containsKey(Object o) { return delegate.containsKey(o); }
+        @Override public int                  size() { return delegate.size(); }
+        @Override public Set<Map.Entry<K, V>> entrySet() { return new LazyMapEntrySet(); }
+
+        @ForceInline
+        @Override
+        public V get(Object key) {
+            final StableValueImpl<V> stable = delegate.get(key);
+            if (stable == null) {
+                return null;
+            }
+            @SuppressWarnings("unchecked")
+            final K k = (K) key;
+            return computeIfUnset(k, stable);
+        }
+
+        @ForceInline
+        V computeIfUnset(K key, StableValueImpl<V> stable) {
+            V v = stable.value();
+            if (v != null) {
+                return StableValueImpl.unwrap(v);
+            }
+            synchronized (stable) {
+                v = stable.value();
+                if (v != null) {
+                    return StableValueImpl.unwrap(v);
+                }
+                v = mapper.apply(key);
+                stable.setOrThrow(v);
+            }
+            return v;
+        }
+
+        @jdk.internal.ValueBased
+        final class LazyMapEntrySet extends AbstractImmutableSet<Map.Entry<K, V>> {
+
+            @Stable
+            private final Set<Map.Entry<K, StableValueImpl<V>>> delegateEntrySet;
+
+            LazyMapEntrySet() {
+                this.delegateEntrySet = delegate.entrySet();
+            }
+
+            @Override public Iterator<Map.Entry<K, V>> iterator() { return new LazyMapIterator(); }
+            @Override public int                       size() { return delegateEntrySet.size(); }
+
+            @Override
+            public int hashCode() {
+                int h = 0;
+                for (Map.Entry<K, V> e : this) {
+                    h += e.hashCode();
+                }
+                return h;
+            }
+
+            @jdk.internal.ValueBased
+            final class LazyMapIterator implements Iterator<Map.Entry<K, V>> {
+
+                @Stable
+                private final Iterator<Map.Entry<K, StableValueImpl<V>>> delegateIterator;
+
+                LazyMapIterator() {
+                    this.delegateIterator = delegateEntrySet.iterator();
+                }
+
+                @Override public boolean hasNext() { return delegateIterator.hasNext(); }
+
+                @Override
+                public Entry<K, V> next() {
+                    final Map.Entry<K, StableValueImpl<V>> inner = delegateIterator.next();
+                    final K key = inner.getKey();
+                    return new KeyValueHolder<>(key, computeIfUnset(key, inner.getValue()));
```

Comment:
This needs to return an `AbstractMap​.SimpleImmutableEntry`, as `mapper.apply(key)` is allowed to return `null`:
```suggestion
                    return new AbstractMap.SimpleImmutableEntry<>(key, computeIfUnset(key, inner.getValue()));
```
or
```suggestion
                    final V value = computeIfUnset(key, inner.getValue()));
                    return value == null
                        ? new AbstractMap.SimpleImmutableEntry<>(key, value)
                        : new KeyValueHolder<>(key, value);
```

Reply:
`jdk.internal.util.NullableKeyValueHolder` would be better, given its non-seralizable nature.